File: pimpam/src/main/java/org/pma/Game.java
package org.pma;

import java.util.ArrayList;
import java.util.List;

public class Game {
	List<DivisibleRule> divisibleRules;
	
	public Game() {
		divisibleRules = new ArrayList<DivisibleRule>();
		divisibleRules.add(new DivisibleRule(3, "Pim")); 
		divisibleRules.add(new DivisibleRule(5, "Pam")); 
	}
	
	public String play(int number) {
		String message = "";
		for (DivisibleRule rule : divisibleRules) {
			message += rule.apply(number);
		}
		return message;
	}
}

File: SCALA_pimpam/src/main/scala/org/pimpam/game.scala
package org.pimpam

object game {
  
	val divisorMessages = (3->"pim", 5->"pam") 

  
  	def play(number: Int) : String = {
  	  applyRule(number, 3, "pim") + applyRule(number, 5, "pam")
	}
  
  	private def applyRule(number: Int, divisor: Int, message: String) : String = {
  	  if (number % divisor == 0) message
  	  else ""
  	}
}